File: 046_Permutations/046_1.java
import java.util.Arrays;
import java.util.Comparator;
import java.util.Deque;
import java.util.ArrayDeque;
import java.util.List;
import java.util.ArrayList;
import java.util.LinkedList;

class Solution {
    List<List<Integer>> res = new LinkedList<>();

    public List<List<Integer>> permute(int[] nums) {
        LinkedList<Integer> track = new LinkedList<>();
        backtrack(nums, track);
        return res;
    }

    private void backtrack(int[] nums, LinkedList<Integer> track) {
        if (track.size() == nums.length) {
            res.add(new LinkedList<>(track));
            return;
        }

        for (int i = 0; i < nums.length; ++i) {
            if (track.contains(nums[i])) continue;
            track.add(nums[i]);
            backtrack(nums, track);
            track.removeLast();
        }
    }
}

public class leetcode {
    public static void main(String[] args) {
        int[] a = {1, 2, 3};
        Solution s = new Solution();
        List<List<Integer>> res = s.permute(a);
        for (List<Integer> i : res) {
            for (Integer j : i)
                System.out.print(j + " ");
            System.out.println();
        }
        // System.out.println(res);
    }
}

File: 046_Permutations/046_1.cpp
#include <iostream>
#include <string>
#include <vector>
#include <climits>
#include <algorithm>
#include <cstring>
#include <unordered_map>
#include <map>
#include <stack>
using namespace std;

class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> res;
        deque<int> track; 
        backtrack(nums, track, res);
        return res;
    }

private:
    void backtrack(vector<int> &nums, deque<int> &track, vector<vector<int>> &res) {
        if (track.size() == nums.size()) {
            res.emplace_back(vector<int>(track.begin(), track.end()));
            return;
        }

        for (int &n : nums) {
            if (find(track.cbegin(), track.cend(), n) != track.end()) continue;
            track.push_back(n);
            backtrack(nums, track, res);
            track.pop_back();
        }
    }
};

int main() {
    vector<int> nums = {1, 2, 3};
    Solution s;
    for (auto & i : s.permute(nums)) {
        for (auto & j : i)
            cout << j << " ";
        cout << "\n";
    }
}
